[WIP] Add IFEval / Instruction-Following Eval

wiskojo opened the pull request:
Add IFEval task #1012.

Results for `gpt-4-1106-preview` from this PR:
```
openai-chat-completions (model=gpt-4-1106-preview), gen_kwargs: (), limit: None, num_fewshot: None, batch_size: 1
|Tasks |Version|Filter|n-shot|        Metric         |Value |   |Stderr|
|------|-------|------|-----:|-----------------------|-----:|---|-----:|
|ifeval|Yaml   |none  |     0|prompt_level_strict_acc|0.7542|±  |0.0185|
|      |       |none  |     0|inst_level_strict_acc  |0.8261|   |      |
|      |       |none  |     0|prompt_level_loose_acc |0.8115|±  |0.0168|
|      |       |none  |     0|inst_level_loose_acc   |0.8681|   |      |
```
Roughly similar to the reported numbers for GPT-4 from the paper:
![Screenshot 2023-12-09 at 12 04 56 AM](https://github.com/EleutherAI/lm-evaluation-harness/assets/22329910/b7619c9b-0ae1-4c75-adee-616da2b99704)

Notes:
- The implementation requires some additional dependencies `langdetect` and `immutabledict` which I added as optional via `pip install -e “.[ifeval]”` in `pyproject.toml` . Not sure if this is a good approach for this.
- The task is configured to generate with `temperature=0` and changed `max_gen_tokens` to 1024. I don't think I saw exact details for generation parameters in the paper are these defaults ok?
- The big issue is the task as implemented only works with `openai-chat-completions` model and doesn't work with e.g. `hf` models. I use `doc_to_text: prompt` but after some testing realized that this doesn't work for chat based models that require specific chat formats e.g. ChatML, Zephyr, Alpaca, etc. and I'm not sure how to make this task compatible across different models. I could e.g. change `doc_to_text: "<|im_start|>user\n{{prompt}}<|im_end|>\n<|im_start|>assistant\n"` but then this would only work for ChatML based `hf` models. Need some help here to identify how to correctly implement this task to work with harness.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/EleutherAI/lm-evaluation-harness?pullRequest=1087) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/EleutherAI/lm-evaluation-harness?pullRequest=1087) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/EleutherAI/lm-evaluation-harness?pullRequest=1087) it.</sub>

haileyschoelkopf commented:
Thanks very much for the contribution, it's much appreciated! Hoping this task will be useful to the community.

- Adding those as optional dependencies should be good. I think `immutabledict` could be dropped without harming results but `langdetect` would still be needed.
> The task is configured to generate with temperature=0 and changed max_gen_tokens to 1024. I don't think I saw exact details for generation parameters in the paper are these defaults ok?
- it's weird the authors don't give more detail on their evaluation setup. These seem good to me! (might need to go a little bit longer? some examples in the dataset mention "900 words" long)
- for now `doc_to_text: prompt` is good, I'll see if I can test this on Claude if possible. For full generality we may want to get #1098 merged, and possibly just make this task print a warning when evaling on a non-chat model in the meantime (or hold off on merging perhaps?)

wiskojo commented:
Ah good catch on the “900 words” long instructions, I totally overlooked that! I’ve bumped `max_gen_tokens` to 2048, that should be more than enough assuming 1 token ~= 3/4 words.

Regarding general compatibility with chat models, definitely need #1098 merged to get the most use of this task to eval different `hf` models. Currently can get this to work by manually modifying `doc_to_text` with the right chat format for whatever model I’m evalling, but that definitely isn’t ideal. Whether we want to merge this PR before or after #1098 I’ll leave to the maintainers discretion. If any modifications should be made before merge, happy to help work on those as needed.

haileyschoelkopf commented:
We can probably get away with `max_gen_tokens: 1280` to ease the cost of using the GPT-4 api on this task / to make local models run faster.

I'd be okay with merging this now with `eval_logger.warning("This task is meant for chat-finetuned models, and may not give meaningful results for models other than `openai` or `anthropic` if `doc_to_text` in its YAML is not wrapped in the appropriate chat template string. This warning will be removed when chat templating support is added natively to local models")` in the `process_results()` call--do you prefer that rather than leaving this PR sitting too long unmerged?

wiskojo commented:
Makes sense, I’ve changed `max_gen_toks: 1280`, added the warning message to the beginning of the `process_results()` call, and fixed formatting issues from the linter.

Not sure how to fix the `test_tasks` test failing though. Looks like the optional dependencies like `langdetect` are not installed before the tests are ran because the installation is done via `pip install -e '.[testing]' --extra-index-url https://download.pytorch.org/whl/cpu`.

haileyschoelkopf commented:
Will see if there's a generalized fix beyond just adding the ifeval extra to the test_tasks install. 

Else, thank you for your work on this, and hoping to look at adding chat templates for HF models this week!

pminervini commented:
Hey, the `ifeval` task is still not visible from `ALL_TASKS` somehow -- what could be the reason for this?

wiskojo commented:
I suspect this might be because the optional dependencies haven't been installed. Have you done `pip install -e “.[ifeval]”`?

lintangsutawika commented:
Also, if you want to debug what the issue might be, you can use `-v DEBUG` when evaluating on `ifeval`
